src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.cs

```diff
@@ -32,68 +33,93 @@ public partial class EventSourceGenerator : ISourceGenerator
         //         }
         //     }
 
-        public void Initialize(GeneratorInitializationContext context)
-            => context.RegisterForSyntaxNotifications(() => new SyntaxReceiver());
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            IncrementalValuesProvider<EventSourceClass> eventSourceClasses =
+                context.SyntaxProvider
+                .CreateSyntaxProvider(static (x, _) => IsSyntaxTargetForGeneration(x), (x, _) => GetSemanticTargetForGeneration(x))
+                .Where(x => x is not null)
+                .Collect()
+                .Combine(context.CompilationProvider)
+                // Use a custom comparer that ignores the compilation.
+                .WithComparer(ClassDeclarationsEqualityComparer.Instance)
+                .SelectMany((x, cancellationToken) => Parser.GetEventSourceClasses(x.Left, x.Right, cancellationToken));
+
+            context.RegisterSourceOutput(eventSourceClasses, static (spc, x) => Emitter.Emit(spc, x));
+        }
 
-        public void Execute(GeneratorExecutionContext context)
+        private static bool IsSyntaxTargetForGeneration(SyntaxNode node) =>
+            node is ClassDeclarationSyntax x && x.AttributeLists.Count > 0;
+
+        private static ClassDeclarationSyntax? GetSemanticTargetForGeneration(GeneratorSyntaxContext context)
         {
-            SyntaxReceiver? receiver = context.SyntaxReceiver as SyntaxReceiver;
-            if ((receiver?.CandidateClasses?.Count ?? 0) == 0)
-            {
-                // nothing to do yet
-                return;
-            }
+            // Only add classes annotated [EventSourceAutoGenerate] to reduce busy work.
+            const string EventSourceAttribute = "EventSourceAutoGenerateAttribute";
```

Comment:
Note that this wont work in the presence of type aliases. That might be an acceptable tradeoff, but something to be aware of.